### PR TITLE
Randomized Terrain Layers

### PR DESCRIPTION
**Pull request by DerCooleVonDem**
Just some little edits on the GroundCover.php Populator that it is able to generate randomized terrain layers. 

I created this populator so that I can generate random terrain layers in my MultiWorld fork. I think you can use this in the future or it is an inspiration/idea :)

### CODE REVIEW

File: src/czechpmdevs/multiworld/generator/normal/NormalGenerator.php [hunk 1]
```diff
@@ -109,8 +109,10 @@ public function init(ChunkManager $level, Random $random): void {
         $this->noiseBase = new Simplex($this->random, 4, 1 / 4, 1 / 32);
         $this->random->setSeed($this->level->getSeed());
         $this->selector = new BiomeSelector($this->random);
-
-        $cover = new GroundCover();
+        
+
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/NormalGenerator.php:113]: Leave only one line

File: src/czechpmdevs/multiworld/generator/normal/NormalGenerator.php [hunk 2]
```diff
@@ -109,8 +109,10 @@ public function init(ChunkManager $level, Random $random): void {
         $this->noiseBase = new Simplex($this->random, 4, 1 / 4, 1 / 32);
         $this->random->setSeed($this->level->getSeed());
         $this->selector = new BiomeSelector($this->random);
-
-        $cover = new GroundCover();
+        
+        
+        //edit the GroundCover Populator that it be able to generate randomized terrain :) (Could be useful in future)
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/NormalGenerator.php:114]: This comment will be useless in the future

File: src/czechpmdevs/multiworld/generator/normal/biome/River.php [hunk 3]
```diff
@@ -29,8 +29,11 @@ class River extends Biome {
 
     public function __construct() {
         parent::__construct(0.5, 0.5);
+        $sand = Block::SAND;
+        $gravel = Block::GRAVEL;
+        //EXAMPLE POPULATION
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/biome/River.php:34]: Useless comment

File: src/czechpmdevs/multiworld/generator/normal/biome/River.php [hunk 4]
```diff
@@ -29,8 +29,11 @@ class River extends Biome {
...
         $this->setGroundCover([
-            new Sand(),
+            ["{$sand}:0", "{$gravel}:0"],
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/biome/River.php:36]: Don't use strings to pass item ids, use array or any other better way

File: src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+use pocketmine\block\BlockFactory;
+use pocketmine\block\Liquid;
+use pocketmine\level\biome\Biome;
+use pocketmine\level\ChunkManager;
+use pocketmine\level\generator\populator\GroundCover;
+use pocketmine\level\generator\populator\Populator;
+use pocketmine\utils\Random;
+
+class GroundCoverCustom extends Populator{
+
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php:35]: 1 line is enough

File: src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+use pocketmine\block\Block;
+use pocketmine\block\BlockFactory;
+use pocketmine\block\Liquid;
+use pocketmine\level\biome\Biome;
+use pocketmine\level\ChunkManager;
+use pocketmine\level\generator\populator\GroundCover;
+use pocketmine\level\generator\populator\Populator;
+use pocketmine\utils\Random;
+
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php:33]: Missing implemented functions

File: src/czechpmdevs/multiworld/generator/normal/biome/River.php [hunk 7]
```diff
@@ -29,8 +29,11 @@ class River extends Biome {
 
     public function __construct() {
         parent::__construct(0.5, 0.5);
+        $sand = Block::SAND;
+        $gravel = Block::GRAVEL;
+        //EXAMPLE POPULATION
         $this->setGroundCover([
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/biome/River.php:35]: Why is it changed only in River biome?

File: src/czechpmdevs/multiworld/generator/normal/biome/River.php [hunk 8]
```diff
@@ -29,8 +29,11 @@ class River extends Biome {
 
     public function __construct() {
         parent::__construct(0.5, 0.5);
+        $sand = Block::SAND;
+        $gravel = Block::GRAVEL;
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/biome/River.php:33]: Useless definitions

File: src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                        $b = $cover[$startY - $y];
+
+
+                        if($column[$y] === "\x00" and $b->isSolid()){
+                            break;
+                        }
+                        if(is_array($b)){
+                            $selected = $random->nextRange(0, count($b)-1);
+                            $randContent = $b[$selected];
+                            $data = explode(":",(string)$randContent);
```
Comment (VixikHD) [inline src/czechpmdevs/multiworld/generator/normal/populator/impl/GroundCoverCustom.php:65]: Don't use strings to pass blocks

**Review by VixikHD — CHANGES_REQUESTED** [review]
Have you tested it?

**Comment by DerCooleVonDem** [comment]
> Have you tested it?

Yes ive tested it and it works as intended

**Comment by DerCooleVonDem** [comment]
I now change everything like you want it to

File: src/czechpmdevs/multiworld/generator/normal/biome/River.php (hunk 7 above)
Comment (DerCooleVonDem) [reply src/czechpmdevs/multiworld/generator/normal/biome/River.php:35]: Its a example to show you my idea

Comment (DerCooleVonDem) [reply src/czechpmdevs/multiworld/generator/normal/biome/River.php:35]: will be changed as soon as possible

**Comment by benda95280** [comment]
Any news ? :)

**Pull request closed without merging**